src/v/archival/archiver_manager.cc

```diff
@@ -0,0 +1,1051 @@
+};
+
+class archiver_manager::impl {
+public:
+    impl(
+      model::node_id node_id,
+      ss::sharded<cluster::partition_manager>& pm,
+      ss::sharded<raft::group_manager>& gm,
+      ss::sharded<cloud_storage::remote>& api,
+      ss::sharded<cloud_storage::cache>& cache,
+      ss::sharded<archival::upload_housekeeping_service>& upload_housekeeping,
+      ss::lw_shared_ptr<const configuration>& config)
+      : _self_node_id(node_id)
+      , _pm(pm)
+      , _gm(gm)
+      , _remote(api)
+      , _cache(cache)
+      , _upload_housekeeping(upload_housekeeping)
+      , _config(std::move(config))
+      , _rtc(_as)
+      , _logger(archival_log, _rtc, ssx::sformat("node-{}", node_id)) {
+        vlog(_logger.info, "Create archiver_manager");
+    }
+
+    ss::future<> start() {
+        vlog(_logger.info, "Start archiver_manager");
+        auto gh = _gate.hold();
+
+        vassert(
+          _pm.local_is_initialized(), "partition_manager is not initialized");
+        vassert(
+          !config::shard_local_cfg()
+             .cloud_storage_disable_archiver_manager.value(),
+          "Legacy mode enabled");
+
+        // Partition movement notification
+        _manage_notifications = _pm.local().register_manage_notification(
+          model::kafka_namespace,
+          [this](ss::lw_shared_ptr<cluster::partition> new_partition) {
+              start_managing_partition(std::move(new_partition));
+          });
+
+        _unmanage_notifications = _pm.local().register_unmanage_notification(
+          model::kafka_namespace, [this](model::topic_partition_view tpv) {
+              model::ntp ntp(model::kafka_namespace, tpv.topic, tpv.partition);
+              stop_managing_partition(ntp);
+          });
+
+        vassert(_gm.local_is_initialized(), "group_manager is not initialized");
+
+        _leadership_notifications
+          = _gm.local().register_leadership_notification(
+            [this](
+              raft::group_id group,
+              model::term_id term,
+              std::optional<model::node_id> leader_id) {
+                notify_leadership_change(group, term, leader_id);
+            });
+        co_return;
+    }
+
+    ss::future<> stop() {
+        vlog(_logger.debug, "Stopping archiver_manager");
+        if (_manage_notifications) {
+            _pm.local().unregister_manage_notification(
+              _manage_notifications.value());
+        }
+        if (_unmanage_notifications) {
+            _pm.local().unregister_unmanage_notification(
+              _unmanage_notifications.value());
+        }
+        if (_leadership_notifications) {
+            _gm.local().unregister_leadership_notification(
+              _leadership_notifications.value());
+        }
+        for (auto& [k, v] : _managed) {
+            vlog(
+              _logger.debug,
+              "Stopping archiver for {}, ref-count: {}",
+              k,
+              v.use_count());
+            co_await v->stop();
```

Comment:
This iteration looks suspicious with the scheduling point. Though would we even get here? Would we expect `_managed` to be clear by the running of the unmanage notifications?

Reply:
It's not modifying the container so should be safe with scheduling point.

Reply:
Isn't it possible that `_managed` has changed because we've serviced `stop_managing_partition()`?

Reply:
notifications are already disabled at this point (line 830)